Toolkit/ArcGISToolkit/TimeSlider.swift

```diff
@@ -0,0 +1,2609 @@
+        fullExtentEndTimeLabel.foregroundColor = fullExtentLabelColor.cgColor
+        fullExtentEndTimeLabel.alignmentMode = kCAAlignmentCenter
+        fullExtentEndTimeLabel.frame = CGRect.zero
+        fullExtentEndTimeLabel.contentsScale = UIScreen.main.scale
+        fullExtentEndTimeLabel.fontSize = fullExtentLabelFont.pointSize
+        layer.addSublayer(fullExtentEndTimeLabel)
+        
+        // Add the lower value label
+        currentExtentStartTimeLabel.isHidden = !(labelMode == .thumbs && isSliderVisible)
+        currentExtentStartTimeLabel.foregroundColor = currentExtentLabelColor.cgColor
+        currentExtentStartTimeLabel.alignmentMode = kCAAlignmentCenter
+        currentExtentStartTimeLabel.frame = CGRect.zero
+        currentExtentStartTimeLabel.contentsScale = UIScreen.main.scale
+        currentExtentStartTimeLabel.fontSize = currentExtentLabelFont.pointSize
+        layer.addSublayer(currentExtentStartTimeLabel)
+        
+        // Add the upper value label
+        currentExtentEndTimeLabel.isHidden = !(labelMode == .thumbs && isSliderVisible)
+        currentExtentEndTimeLabel.foregroundColor = currentExtentLabelColor.cgColor
+        currentExtentEndTimeLabel.alignmentMode = kCAAlignmentCenter
+        currentExtentEndTimeLabel.frame = CGRect.zero
+        currentExtentEndTimeLabel.contentsScale = UIScreen.main.scale
+        currentExtentEndTimeLabel.fontSize = currentExtentLabelFont.pointSize
+        layer.addSublayer(currentExtentEndTimeLabel)
+        
+        // Create the images
+        let bundle = Bundle(for: type(of: self))
+        playImage = UIImage(named: "Play", in: bundle, compatibleWith: nil)
+        pauseImage = UIImage(named: "Pause", in: bundle, compatibleWith: nil)
+        forwardImage = UIImage(named: "Forward", in: bundle, compatibleWith: nil)
+        backImage = UIImage(named: "Back", in: bundle, compatibleWith: nil)
+        
+        // Add Play/Pause button
+        let p = UIButton(type: .custom)
+        p.setBackgroundImage(playImage, for: .normal)
+        p.addTarget(self, action: #selector(TimeSlider.playPauseAction(_:)), for: .touchUpInside)
+        p.showsTouchWhenHighlighted = true
+        addSubview(p)
+        playPauseButton = p
+        
+        // Add forward button
+        let f = UIButton(type: .custom)
+        f.setBackgroundImage(forwardImage, for: .normal)
+        f.addTarget(self, action: #selector(TimeSlider.forwardAction(_:)), for: .touchUpInside)
+        f.showsTouchWhenHighlighted = true
+        addSubview(f)
+        forwardButton = f
+        
+        // Add back button
+        let b = UIButton(type: .custom)
+        b.setBackgroundImage(backImage, for: .normal)
+        b.addTarget(self, action: #selector(TimeSlider.backAction(_:)), for: .touchUpInside)
+        b.showsTouchWhenHighlighted = true
+        addSubview(b)
+        backButton = b
+        
+        // Refresh
+        refresh()
+    }
+    
+    private func refresh() {
```

Comment:
It seems to me that the contents of this function could be inlined into `layoutSubviews()`. And whenever a property changes that would affect layout, `setNeedsLayout()` should be called instead. That would prevent layout from happening more than once when multiple properties change.

Reply:
done